crates/query-engine/translation/src/translation/query/values.rs

```diff
@@ -84,48 +84,88 @@ fn type_to_ast_scalar_type(typ: &database::Type) -> sql::ast::ScalarType {
-            table: variables_table,
-            column: sql::helpers::make_column_alias(sql::helpers::VARIABLES_FIELD.to_string()),
-        });
+    let variables_reference = Expression::ColumnReference(ColumnReference::AliasedColumn {
+        table: variables_table,
+        column: sql::helpers::make_column_alias(sql::helpers::VARIABLES_FIELD.to_string()),
+    });
 
-    match r#type {
-        database::Type::CompositeType(_type_name) => {
-            let exp = sql::ast::Expression::BinaryOperation {
+    let projected_variable_exp = match r#type {
+        database::Type::CompositeType(_) | database::Type::ArrayType(_) => {
+            sql::ast::Expression::BinaryOperation {
                 left: Box::new(variables_reference),
                 operator: sql::ast::BinaryOperator("->".to_string()),
                 right: Box::new(sql::ast::Expression::Value(sql::ast::Value::String(
                     variable.clone(),
                 ))),
-            };
-            sql::ast::Expression::FunctionCall {
-                function: sql::ast::Function::JsonbPopulateRecord,
-                args: vec![
-                    sql::ast::Expression::Cast {
-                        expression: Box::new(sql::ast::Expression::Value(sql::ast::Value::Null)),
-                        r#type: type_to_ast_scalar_type(r#type),
-                    },
-                    exp,
-                ],
             }
         }
-        _ => {
-            let exp = sql::ast::Expression::BinaryOperation {
-                left: Box::new(variables_reference),
-                operator: sql::ast::BinaryOperator("->>".to_string()),
-                right: Box::new(sql::ast::Expression::Value(sql::ast::Value::String(
-                    variable.clone(),
-                ))),
+        database::Type::ScalarType(_) => sql::ast::Expression::BinaryOperation {
+            left: Box::new(variables_reference),
+            operator: sql::ast::BinaryOperator("->>".to_string()),
+            right: Box::new(sql::ast::Expression::Value(sql::ast::Value::String(
+                variable.clone(),
+            ))),
+        },
+    };
+
+    translate_projected_variable(state, r#type, projected_variable_exp)
+}
+
+pub fn translate_projected_variable(
+    state: &mut State,
+    r#type: &database::Type,
+    exp: sql::ast::Expression,
+) -> sql::ast::Expression {
+    match r#type {
+        database::Type::CompositeType(_type_name) => sql::ast::Expression::FunctionCall {
+            function: sql::ast::Function::JsonbPopulateRecord,
+            args: vec![
+                sql::ast::Expression::Cast {
+                    expression: Box::new(sql::ast::Expression::Value(sql::ast::Value::Null)),
+                    r#type: type_to_ast_scalar_type(r#type),
+                },
+                exp,
+            ],
+        },
+        database::Type::ArrayType(type_name) => {
+            let arr_table = state.make_table_alias("arr".to_string());
```

Comment:
nit: can we please avoid abbreviations in code and names? lets use:
- `let array_table`
- `"array".to_string`
- `let element_column`
- `name: "element".to_string`
- `expression: `
- `element_expression`